File: project/dexter-monitor/routes/defect.js
/**
 * Copyright (c) 2016 Samsung Electronics, Inc.,
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
"use strict";

const mysql = require("mysql");
const moment = require('moment');
const Promise = require('bluebird');
const rp = require('request-promise');
const _ = require("lodash");
const project = require('./project');
const server = require('./server');
const database = require("../util/database");
const route = require('./route');
const log = require('../util/logging');

exports.getAll = function(req, res) {
    const sql =
        "SELECT year, week, groupName, projectName, language, allDefectCount,   "+
        "       allNew, allFix, allDis, criNew, criFix, criDis,                 "+
        "       majNew, majFix, majDis, minNew, minFix, minDis,                 "+
        "       crcNew, crcFix, crcDis, etcNew, etcFix, etcDis                  "+
        "FROM WeeklyStatus                                                      "+
        "LEFT JOIN ProjectInfo                                                  "+
        "ON WeeklyStatus.pid = ProjectInfo.pid                                  "+
        "ORDER BY year DESC, week DESC,                                         "+
        "         groupName ASC, projectName ASC                                ";

    route.executeSqlAndSendResponseRows(sql, res);
};

exports.getByGroup = function(req, res) {
    const year = mysql.escape(req.params.year);
    const week = mysql.escape(req.params.week);

    let sql =
        "SELECT year, week, groupName,                                  " +
        "       SUM(userCount) AS userCount,                            " +
        "       COUNT(projectName) AS projectCount,                     " +
        "       SUM(allDefectCount) AS allDefectCount,                  " +
        "       SUM(allFix) AS allFix,                                  " +
        "       SUM(allDis) AS allDis                                   " +
        "FROM WeeklyStatus                                              " +
        "LEFT JOIN ProjectInfo                                          " +
        "ON WeeklyStatus.pid = ProjectInfo.pid                          ";

    if (year == 0 && week == 0) {
        sql += "WHERE year = YEAR(CURDATE()) AND week = WEEK(CURDATE()) ";
    } else {
        sql += "WHERE year = " + year + " AND week = " + week + "       ";
    }

    sql += "GROUP BY groupName ORDER BY allDefectCount DESC, groupName ASC";

    route.executeSqlAndSendResponseRows(sql, res);
};

exports.getMinYear = function(req, res) {
    const sql = "SELECT year FROM WeeklyStatus ORDER BY year ASC LIMIT 1";
    database.exec(sql)
        .then((rows) => {
            res.send({status:'ok', value: rows[0].year});
        })
        .catch((err) => {
            log.error(err);
            res.send({status:"fail", errorMessage: err.message});
        });
};

exports.getMaxYear = function(req, res) {
    const sql = "SELECT year FROM WeeklyStatus ORDER BY year DESC LIMIT 1";
    database.exec(sql)
        .then((rows) => {
            res.send({status:'ok', value: rows[0].year});
        })
        .catch((err) => {
            log.error(err);
            res.send({status:"fail", errorMessage: err.message});
        });
};

exports.getMaxWeek = function(req, res) {
    const year = mysql.escape(req.params.year);
    const sql = "SELECT week FROM WeeklyStatus WHERE year = " + year + " ORDER BY week DESC LIMIT 1";
    database.exec(sql)
        .then((rows) => {
            res.send({status:'ok', value: rows[0].week});
        })
        .catch((err) => {
            log.error(err);
            res.send({status:"fail", errorMessage: err.message});
        });
};

exports.getDefectCountByProjectName = function(req, res) {
    const projectServer = _.find(server.getServerListInternal(), {'projectName': req.params.projectName});
    const defectCountUrl = `http://${projectServer.hostIP}:${projectServer.portNumber}/api/v2/defect-count`;
    rp(defectCountUrl)
        .then((data) => {
            const parsedData = JSON.parse('' + data);
            res.send({status:'ok', values: parsedData.values});
        })
        .catch((err) => {
            log.error(`Failed to get defect count for pid ${projectServer.pid} : ${err}`);
            res.send({status:"fail", errorMessage: err.message});
        });
};

exports.saveSnapshot = function() {
    const activeServerList = _.filter(server.getServerListInternal(), {'active': true});

    Promise.map(activeServerList, (server) => {
        const defectCountUrl = `http://${server.hostIP}:${server.portNumber}/api/v2/detailed-defect-count`;
        const userCountUrl = `http://${server.hostIP}:${server.portNumber}/api/v2/user-count`;
        let promises = [];
        let defectValues = {};
        let userCount = 0;

        promises.push(new Promise((resolve, reject) => {
            rp(defectCountUrl)
                .then((data) => {
                    defectValues = JSON.parse('' + data).values;
                    resolve();
                })
                .catch((err) => {
                    log.error(`Failed to get detailed defect count for pid ${server.pid} : ${err}`);
                    reject();
                });
        }));
        promises.push(new Promise((resolve, reject) => {
            rp(userCountUrl)
                .then((data) => {
                    userCount = JSON.parse('' + data).value;
                    resolve();
                })
                .catch((err) => {
                    log.error(`Failed to get user count for pid ${server.pid} : ${err}`);
                    reject();
                });
        }));

        Promise.all(promises)
            .then(() => {
                insertSnapshotToDatabase(server.pid, defectValues, userCount);
            })
            .catch((err) => {
                log.error(err);
            });
    }).catch((err) => {
        log.error(err);
    });
};

function insertSnapshotToDatabase(pid, defectValues, userCount) {
    const year = moment().get('year');
    const weekOfYear = moment().isoWeek();
    const dayOfWeek = moment().isoWeekday();
    const {allDefectCount, allNew, allFix, allDis, criNew, criFix, criDis,
        majNew, majFix, majDis, minNew, minFix, minDis,
        crcNew, crcFix, crcDis, etcNew, etcFix, etcDis,
        invalidStatusCode, invalidSeverityCode} = defectValues;

    const sql =
        `INSERT INTO WeeklyStatus(pid, year, week, day, userCount, allDefectCount,
                                    allNew, allFix, allDis, criNew, criFix, criDis,
                                    majNew, majFix, majDis, minNew, minFix, minDis,
                                    crcNew, crcFix, crcDis, etcNew, etcFix, etcDis)
                     VALUES (${pid}, ${year}, ${weekOfYear}, ${dayOfWeek}, ${userCount}, ${allDefectCount},
                            ${allNew}, ${allFix}, ${allDis}, ${criNew}, ${criFix}, ${criDis},
                            ${majNew}, ${majFix}, ${majDis}, ${minNew}, ${minFix}, ${minDis},
                            ${crcNew}, ${crcFix}, ${crcDis}, ${etcNew}, ${etcFix}, ${etcDis})`;

    database.exec(sql)
        .then(() => {
            log.info(`Inserted snapshot for pid ${pid} : # of invalidStatusCode: ${invalidStatusCode}, # of invalidSeverityCode: ${invalidSeverityCode}`);

        })
        .catch((err) => {
            log.error(`Failed to insert snapshot for pid ${pid} : ${err}`);
        });
}

File: project/dexter-core/src/java/com/samsung/sec/dexter/core/defect/Defect.java
/**
 * Copyright (c) 2014 Samsung Electronics, Inc.,
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 * 
 * * Redistributions of source code must retain the above copyright notice, this
 * list of conditions and the following disclaimer.
 * 
 * * Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
package com.samsung.sec.dexter.core.defect;

import com.google.common.base.Objects;
import com.google.gson.Gson;
import com.samsung.sec.dexter.core.BaseDefect;
import com.samsung.sec.dexter.core.analyzer.ResultFileConstant;
import com.samsung.sec.dexter.core.exception.DexterException;
import com.samsung.sec.dexter.core.filter.DefectFilter;
import com.samsung.sec.dexter.core.util.DexterUtil;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

public class Defect extends BaseDefect {
    /** set by server module not by local */
    private long gdid = -1;
    private transient String localDid = "" + System.currentTimeMillis();
    private List<Occurence> occurences = new ArrayList<Occurence>();
    private long createdDateTime = -1;
    private String message = "";
    private String severityCode = "";
    private String categoryName = "";
    private long modifiedDateTime = -1;
    private String analysisType = "";

    public Defect() {
        createdDateTime = System.currentTimeMillis();
        modifiedDateTime = createdDateTime;
    }

    /*
     * (non-Javadoc)
     * 
     * @see java.lang.Object#toString()
     */
    @Override
    public String toString() {
        return Objects.toStringHelper(this).add("did", this.gdid)
                .add(checkerCode, checkerCode)
                .add(ResultFileConstant.TOOL_NAME, toolName)
                .add(ResultFileConstant.LANGUAGE, language)
                .add(ResultFileConstant.CLASS_NAME, className)
                .add(ResultFileConstant.FILE_NAME, getFileName())
                .add(ResultFileConstant.MODULE_PATH, getModulePath())
                .add("occ size", occurences.size())
                .add("AnalysisType", getAnalysisType())
                .add(ResultFileConstant.OCCURENCES, occurences).toString();
    }

    /*
     * (non-Javadoc)
     * 
     * @see java.lang.Object#equals(java.lang.Object)
     */
    @Override
    public boolean equals(final Object obj) {
        return super.equals(obj);
    }

    public boolean equalsWithPreOccurence(final Object obj) {
        if (obj == null) {
            return false;
        }

        return super.equals(obj);
    }

    /*
     * (non-Javadoc)
     * 
     * @see java.lang.Object#hashCode()
     */
    @Override
    public int hashCode() {
        return super.hashCode();
    }

    public String toJson() {
        Gson gson = new Gson();
        return gson.toJson(this);
    }

    public void addOccurence(final Occurence occurence) {
        assert this.occurences.contains(occurence) == false;

        String key = "[#" + (occurences.size() + 1) + "@" + occurence.getStartLine() + "] ";
        //		if(this.message.indexOf(key) < 0){
        //		}
        this.message += key + occurence.getMessage() + " ";
        occurences.add(occurence);
    }

    /**
     * @return Object
     */
    public Occurence getFirstOccurence() {
        final Iterator<Occurence> iter = this.occurences.iterator();
        if (iter.hasNext()) {
            return iter.next();
        }

        return null;
    }

    /**
     * @return the occurences
     */
    public List<Occurence> getOccurences() {
        return occurences;
    }

    /**
     * @param occurences
     * the occurences to set
     */
    public void setOccurences(final List<Occurence> occurences) {
        this.occurences = occurences;
    }

    /**
     * @return the localDid
     */
    public String getLocalDid() {
        return localDid;
    }

    /**
     * @return the createdDateTime
     */
    public long getCreatedDateTime() {
        return createdDateTime;
    }

    /**
     * @return the message
     */
    public String getMessage() {
        return message;
    }

    /**
     * @return the lastModifiedTime
     */
    public long getLastModifiedTime() {
        return getModifiedDateTime();
    }

    /**
     * @return the modifiedDateTime
     */
    public long getModifiedDateTime() {
        return modifiedDateTime;
    }

    /**
     * @param createdDateTime
     * the createdDateTime to set
     */
    public void setCreatedDateTime(final long createdTime) {
        this.createdDateTime = createdTime;
    }

    /**
     * please use this carefully, because this field can be set automatically, when adding a occurrence.
     * 
     * @param message
     * the message to set
     */
    public void setMessage(final String message) {
        this.message = message;
    }

    /**
     * @param lastModifiedTime
     * the lastModifiedTime to set
     */
    public void setLastModifiedTime(final long lastModifiedTime) {
        setModifiedDateTime(lastModifiedTime);
    }

    /**
     * @param modifiedDateTime
     * the modifiedDateTime to set
     */
    public void setModifiedDateTime(final long lastModifiedTime) {
        this.modifiedDateTime = lastModifiedTime;
    }

    /**
     * @return the severity
     */
    public String getSeverityCode() {
        return this.severityCode;
    }

    /**
     * @param severity
     * the severity to set
     */
    public void setSeverityCode(final String severityCode) {
        this.severityCode = severityCode;
    }

    /**
     * @return the category
     */
    public String getCategoryName() {
        return this.categoryName;
    }

    /**
     * @param category
     * the severity to set
     */

    public void setCategoryName(final String categoryName) {
        this.categoryName = categoryName;
    }

    /**
     * @param currentUserNo
     * @return DefectFilter
     * @throws DexterException
     */
    public DefectFilter toDefectFilter() {
        final DefectFilter filter = new DefectFilter();
        filter.setActive(true);
        filter.setCheckerCode(checkerCode);
        filter.setClassName(className);
        filter.setCreatedDateTime(System.currentTimeMillis());

        filter.setToolName(toolName);
        filter.setLanguage(language);
        filter.setFileName(getFileName());
        filter.setMethodName(methodName);
        filter.setModulePath(getModulePath());

        return filter;
    }

    /**
     * @return the gdid
     */
    public long getGdid() {
        return gdid;
    }

    /**
     * @param gdid
     * the gdid to set
     */
    public void setGdid(final long gdid) {
        this.gdid = gdid;
    }

    /**
     * @param string
     * @return
     */
    public static Defect fromMap(final Map<String, Object> map) {
        final Defect defect = new Defect();
        defect.setCreatedDateTime(Double.doubleToLongBits((Double) map.get("createdDateTime")) / 1000);
        //defect.setMessage((String) map.get("message"));
        defect.setSeverityCode((String) map.get(ResultFileConstant.SEVERITY_CODE));
        defect.setCategoryName((String) map.get(ResultFileConstant.CATEGORY_NAME));

        defect.setModifiedDateTime(Double.doubleToLongBits((Double) map.get("modifiedDateTime")) / 1000);
        defect.setCheckerCode((String) map.get(ResultFileConstant.CHECKER_CODE));
        defect.setMethodName((String) map.get(ResultFileConstant.METHOD_NAME));
        defect.setToolName((String) map.get(ResultFileConstant.TOOL_NAME));
        defect.setLanguage((String) map.get(ResultFileConstant.LANGUAGE));
        defect.setFileName((String) map.get(ResultFileConstant.FILE_NAME));
        defect.setModulePath((String) map.get(ResultFileConstant.MODULE_PATH));
        defect.setClassName((String) map.get(ResultFileConstant.CLASS_NAME));

        @SuppressWarnings("unchecked")
        final List<Map<String, Object>> occMap = (List<Map<String, Object>>) map.get(ResultFileConstant.OCCURENCES);
        for (final Map<String, Object> om : occMap) {
            Occurence o = new Occurence();

            o.setCode((String) om.get("code"));
            o.setStartLine(DexterUtil.toInt((Double) om.get(ResultFileConstant.START_LINE)));
            o.setEndLine(DexterUtil.toInt((Double) om.get(ResultFileConstant.END_LINE)));
            o.setCharStart(DexterUtil.toInt((Double) om.get(ResultFileConstant.CHAR_START)));
            o.setCharEnd(DexterUtil.toInt((Double) om.get(ResultFileConstant.CHAR_END)));
            o.setVariableName((String) om.get(ResultFileConstant.VARIABLE_NAME));
            o.setStringValue((String) om.get(ResultFileConstant.STRING_VALUE));
            o.setFieldName((String) om.get(ResultFileConstant.FIELD_NAME));
            o.setMessage((String) om.get(ResultFileConstant.MESSAGE));

            defect.addOccurence(o);
        }

        return defect;
    }

    public String getOccurenceLines(String separator) {
        int size = this.occurences.size() * 500;
        StringBuilder lines;

        if (size < Integer.MAX_VALUE)
            lines = new StringBuilder(this.occurences.size() * 500);
        else
            lines = new StringBuilder(Integer.MAX_VALUE);

        for (Occurence o : this.occurences) {
            lines.append(o.getStartLine()).append(separator);
        }

        return lines.toString();
    }

    public String getShortDescription() {
        StringBuilder msg = new StringBuilder(1024);
        msg.append(this.checkerCode)
                .append(" / ").append(this.severityCode)
                .append(" (count:").append(this.occurences.size()).append(", ")
                .append(" lines: ").append(getOccurenceLines(" ")).append(")").append(this.message);
        return msg.toString();
    }

    public String getAnalysisType() {
        return analysisType;
    }

    public void setAnalysisType(String analysisType) {
        this.analysisType = analysisType;
    }
}
